lib/rules/require-valid-css-selector-in-test-helpers.js

```diff
@@ -0,0 +1,158 @@
+    return false;
+  }
+
+  return calleeName[0] === assertIdentifierName;
+}
+
+function _isQuerySelectorCall(node, isTestCall) {
+  const calleeName = parseCallee(node);
+
+  // only validate querySelector calls in the test context
+  if (!isTestCall) {
+    return false;
+  }
+
+  return QUERY_SELECTOR_METHODS.has(calleeName[1]) && PARENT_NODE_NAMES.has(calleeName[0]);
+}
+
+function _isTestHelperCall(node, hasTestHelperImport, localImportNames) {
+  const calleeName = parseCallee(node) || [];
+
+  return hasTestHelperImport && localImportNames.includes(calleeName[0]);
+}
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'disallow using invalid CSS selectors in test helpers',
+      category: 'Testing',
+      recommended: false,
+      url:
+        'https://github.com/ember-cli/eslint-plugin-ember/tree/master/docs/rules/require-valid-css-selector-in-test-helpers.md',
+    },
+    fixable: 'code',
+    schema: [],
+    messages: Object.keys(SELECTOR_RULES).reduce((accumulator, currentVal) => {
+      const _accumulator = accumulator || {};
+      _accumulator[currentVal] = SELECTOR_RULES[currentVal].errorMessage;
+
+      return _accumulator;
+    }, null),
+  },
+  create(context) {
+    let hasTestHelperImport = false;
+    let localImportNames = [];
+    let testImportLocalName = '';
+    let isTestCall = false;
+    let assertIdentifierName = '';
+
+    return {
+      'ImportDeclaration[source.value="@ember/test-helpers"]'(node) {
+        hasTestHelperImport = getPropertyValue(node, 'specifiers').find((specifier) =>
+          TEST_HELPER_IMPORTS.has(getPropertyValue(specifier, 'imported.name'))
+        );
+
+        localImportNames = getPropertyValue(node, 'specifiers')
+          .filter((specifier) =>
+            TEST_HELPER_IMPORTS.has(getPropertyValue(specifier, 'imported.name'))
+          )
+          .map((specifier) => getPropertyValue(specifier, 'local.name'));
+      },
+      'ImportDeclaration[source.value="qunit"]'(node) {
+        const testImportSpecifier = getPropertyValue(node, 'specifiers').find(
+          (specifier) => getPropertyValue(specifier, 'imported.name') === 'test'
+        );
+
+        if (testImportSpecifier) {
+          testImportLocalName = getPropertyValue(testImportSpecifier, 'local.name');
+        }
+      },
+      CallExpression(node) {
+        if (node.callee.name === testImportLocalName) {
+          isTestCall = true;
+          const [, testFn] = node.arguments || [];
+
+          if (testFn) {
+            assertIdentifierName = getPropertyValue(testFn, 'params.0.name');
+          }
+        }
+
+        const value = getPropertyValue(node, 'arguments.0.value');
+
+        if (
+          typeof value !== 'string' ||
+          (!_isAssertDomCall(node, assertIdentifierName) &&
+            !_isTestHelperCall(node, hasTestHelperImport, localImportNames) &&
+            !_isQuerySelectorCall(node, isTestCall))
```

Comment:
I think the current implementation would only flag `this.element.querySelector('[data-derp')` inside `test(...)`, but we need to also catch invalid usages within `beforeEach`/`afterEach` (for both QUnit and Mocha).

Can you add some "bad" test cases for that also?

Reply:
Done